### ISSUE TITLE
RPCs can sometimes call the wrong function

### ISSUE THREAD

**Issue by CloverRoss** [godotengine/godot#95753]
### Tested versions

- Reproducible in 4.2.2.stable.official and 4.3.stable.official (latest as of this date).

### System information

Mac Sonoma 14.1.2

### Issue description
[edit] - fixed post formatting.

I have two different godot projects, a server project and a client project that both talk to each other via RPCs. The server is meant to function as a dedicated server and expects multiple clients. Currently, clients are able to connect to the server and send chat messages to each other, no problems at all (keep this in mind as this is the 'null' case of this issue).
I have added a new RPC to the server and client pair that is completely unrelated to the connection / disconnection process and this causes unexpected behavior in the engine's RPC handling. 

Here is the null case (i.e. the circumstance where things work as expected, before I actually made problematic changes).

Server side:
```gdscript
var time_elapsed = 0
var online_clients = []
var peer = ENetMultiplayerPeer.new()
var version_number = "0.0.0.5"

func _ready():
	print("Server version ",version_number)
	peer.create_server(9281,256)
	multiplayer.multiplayer_peer = peer
	multiplayer.peer_connected.connect(entering)
	multiplayer.peer_disconnected.connect(exiting_server)
	print("Server ready and listening at port ",9281)

@rpc ("any_peer")
func validate_version(id,username):
	rpc_id(id,"validate_version",version_number)
	print("AA")
	server_broadcast(str(username)+" is connecting to the server...")

@rpc ("any_peer")
func user_connection_event(username,joining):
	if joining:
		print("peer joining the server")
		print("BB")
		server_broadcast(username + " has joined the server.")
	else:
		print("CC")
		print("peer leaving the server")
		server_broadcast(username+ " has left the server.")

func exiting_server(x):
	print("leave request from... ",x)
	print("remaining peers: ",multiplayer.get_peers())
	
	
@rpc ("any_peer")
func broadcast(message :String, sender: String):
	print("Broadcast")
	print(message, " ",sender)
	if rpc("broadcast",message,sender) != OK:
		print("Failed broadcast. Server error.")

@rpc("authority")
func server_broadcast(msg :String):
	print("broadcasting server message: ",msg)
	print("DD")
	rpc("server_broadcast", msg)

@rpc ("call_remote")
func entering(x):
	print("received entry request from... ",x)
	rpc_id(x,"entering",x)
	print("online peers: ",multiplayer.get_peers())
```
--
Client side:
```gdscript
var peer = ENetMultiplayerPeer.new()
var localhost = "127.0.0.1" 
var server = "" # redacted for security. not required to repro this issue.
var version_number = "0.0.0.5"
var server_vn = ""
var connection_code = 0
var my_peer_id = 0
var username = "Anon"
var original_username = "Anon"
var first_opened = true 

func requesting_validation():
	rpc_id(1,"validate_version",my_peer_id,username)
	print("---sending validation for ",my_peer_id)

@rpc ("authority")
func validate_version(vn):
	print("---received validation prompt")
	server_vn = vn
	if vn != version_number:
		print("incompatible versions")
		system_validator("/version_incompatible")
	else:
		connection_code = 1
		print("connection code 1")
		system_validator("/serverconnected")

func server_disconnected():
	connection_code = 0
	print("connection code 0")
	system_validator("/serverdisconnected")

func _ready():
	EntryPanel.show()
	MainPanel.hide()
	VersionLabel.text = "Version Number: "+ version_number
	ChangeLog.text = recent_changes
	multiplayer.server_disconnected.connect(server_disconnected)
	#multiplayer.connected_to_server.connect(server_connected)

func connect_to_server():
	OutputText.text = ""
	print("---connecting to server")
	if dev_mode:
		if peer.create_client(localhost,9281) == OK:
			multiplayer.multiplayer_peer = peer
			system_validator("/pass")
		else:
			system_validator("/fail")
	else:
		if peer.create_client(server,9281) == OK:
			multiplayer.multiplayer_peer = peer
			system_validator("/pass")
		else:
			system_validator("/fail")

func disconnect_from_server():
	user_connection_event(username,false) # tell the server we're leaving
	await get_tree().create_timer(0.1).timeout #give enough time for the server to realize we're leaving
	username = original_username
	multiplayer.multiplayer_peer.disconnect_peer(1)

@rpc ("call_remote")
func entering(id):
	print("Entering server")
	my_peer_id = id
	requesting_validation()

@rpc ("call_remote")
func server_broadcast(message :String):
	console(message, "SERVER BROADCAST")

@rpc ("authority")
func user_connection_event(user:String,joining:bool):
	print("user connection event detected.")
	rpc_id(1,"user_connection_event",user,joining)

@rpc("call_remote")
func broadcast(message:String,sender:String):
	print("receiving remote call: ["+message+"] from: "+sender)
	chat(message,sender)

func system_validator(prompt):
	var output = ""
	match prompt:
		"/serverdisconnected": output = "server disconnected."
		"/serverconnected": user_connection_event(username,true)
		"/version_incompatible": 
			output = "
			ERROR: server version number is " + server_vn +" 
			client version number is: " + version_number + "
			Please match version numbers with server or speak with the developer.
			Aborting connection..."
			disconnect_from_server()
		"/ready": output = "welcome, " + username + ". Type /connect to get started. Type /help for more information."
		"/1": output = "Server is listening."
		"/pass": output = "client exists."
		"/fail": output = "client failed. 400"
		"/alerttest": 
			output = "Testing alert."
			alert("This is a test alert.")
		_:
			print("warning: invalid system validation prompt.")
	if output != "":
		print(output)
		console(output,"SYSTEM")

enum SPECIAL_LISTENERS {NONE, RENAME}
var listener = SPECIAL_LISTENERS.NONE
func console(command: String,sender = ""):
	var output = ""
	var clear = false
	var listener_trigger = false
	#print(command)
	match listener:
		SPECIAL_LISTENERS.RENAME:
			if command.begins_with("/cancel"):
				output = "Action cancelled."
			elif command.begins_with("/"):
				output = "Usernames may not begin with a /. Type /rename to try again."
			else:
				username = command
				output = "Username changed to: " + username
		_:
			match command:
				"/clear": 
					clear = true
					OutputText.text = ""
				"/checktime":
					print("Accidental checktime")
					#rpc_id(1,"check_time_elapsed")
				"/connect": 
					output = "connecting to server..."
					connect_to_server()
				"/disconnect":
					output = "Disconnecting from server..."
					disconnect_from_server()
				"/help": output = help_output()
				"/rename": 
					listener_trigger = true
					listener = SPECIAL_LISTENERS.RENAME
					output = "Type in your new username. Type /cancel to keep your current username."
				"/quit":
					get_tree().quit()
				"/q":
					get_tree().quit()
				_: # by default, the console assumes a 'command' is an attempt to broadcast a chat message
					if sender == "":
						if connection_code != 0:  
							var error = rpc_id(1,"broadcast",command,username+" ["+str(my_peer_id)+"]") 
							if error != OK: 
								output = "ERROR: broadcast attempt failed. Code: AA" + str(error)
						else:
							output = "You are not connected."
						print("broadcasting [",command,"] from: ",username)
					else:
						OutputText.text += sender + ": " + command + "\n"
	InputText.clear()
	if !listener_trigger:
		listener = SPECIAL_LISTENERS.NONE
	if !clear && output != "" && sender == "":
		OutputText.text += output + "\n"
	print(output)
```
----
So everything above works as expected - clients can send the '/connect' command into a textinput field, which sets of the server connection process. The server is able to successfully connect with the peer and the client can send broadcasts. This is the null case.

I wanted to add some passage of time to the server code, so I added these lines to the server side:
```gdscript

func _on_timer_timeout():
	ticking_time()
	$Timer.start(1) # should be implied but there is also now a new Timer node on the server side.

func ticking_time():
	time_elapsed += 1

@rpc ("authority")
func check_time_elapsed():
	print("time elapsed")
	pass
	server_broadcast("Time elapsed: "+str(time_elapsed)+" seconds.")
```

When the server and client now try to talk to each other, an error is thrown at the server's "entering" function. When this is called, the line that says 'rpc_id(x,"entering",x)' appears to instead be calling a different method, so the client will throw an error that I'm passing an integer into the 'server_broadcast' method which expects a string. 

Oddly, the troublemaker appears to be the '@rpc ("authority")' line right above my 'func check_time_elapsed():' method. When this is commented out, the null case occurs- it works as expected. The other new methods around the Timer's behavior also don't cause a problem. But specifically the fact I'd named 'func check_time_elapsed():' appears to cause this error.

### Steps to reproduce

I described the repro steps above as it is not a simple case. But essentially:
- set up a separate server and client project which uses rpcs to broadcast text messages to each other
- add a new rpc to the server side 
- see error

### Minimal reproduction project (MRP)

[project.godot.zip](https://github.com/user-attachments/files/16648850/project.godot.zip)
[clientsideproject.godot.zip](https://github.com/user-attachments/files/16648851/clientsideproject.godot.zip)

**Comment by wgeraghty** [comment]
I'm running into this issue in `4.4dev2`.   `4.3` seems fine.  Here's some minimal code to reproduce:

```gdscript
extends Control

@rpc("authority")
func rpc_test_authority(message: String) -> void:
  print("[%d] rpc_test_authority invoked from %d: %s" % [
    multiplayer.get_unique_id(),
    multiplayer.get_remote_sender_id(),
    message
  ])

@rpc("any_peer")
func rpc_test_any_peer(message: String) -> void:
  print("[%d] rpc_test_any_peer invoked from %d:  %s" % [
    multiplayer.get_unique_id(),
    multiplayer.get_remote_sender_id(),
    message
  ])

func _on_peer_connected(id: int) -> void:
  print("[%d] Peer connected: %d" % [
    multiplayer.get_unique_id(),
    id
  ])

func _on_join_pressed() -> void:
  var peer = ENetMultiplayerPeer.new()
  peer.create_client("localhost", 8123)
  multiplayer.multiplayer_peer = peer
  multiplayer.peer_connected.connect(_on_peer_connected)
  print("Joined to localhost:8123")

func _on_host_pressed() -> void:
  var peer = ENetMultiplayerPeer.new()
  peer.create_server(8123)
  multiplayer.multiplayer_peer = peer
  multiplayer.peer_connected.connect(_on_peer_connected)
  print("Hosted on port 8123")

func _on_send_pressed() -> void:
  if multiplayer.is_server():
    print("Calling rpc_test_authority from server")
    rpc_test_authority.rpc("testing rpc_test_authority from server")
  else:
    print("Calling rpc_test_any_peer from client")
    rpc_test_any_peer.rpc("testing rpc_test_any_peer from client")
```

- `Host` on one instance
- `Join` on another instance
- Try `Send` on both

The server invokes the wrong rpc and output shows:

```
[1194654078] rpc_test_any_peer invoked from 1:  testing rpc_test_authority from server
```

The client invokes the wrong rpc and errors with:

```
E 0:00:09:0345   _process_rpc: RPC 'rpc_test_authority' is not allowed on node /root/Control from: 1194654078. Mode is 2, authority is 1.
  <C++ Error>    Condition "!can_call" is true.
  <C++ Source>   modules/multiplayer/scene_rpc_interface.cpp:252 @ _process_rpc()
```

**Comment by Faless** [comment]
> I'm running into this issue in `4.4dev2`

That was a regression in 4.4-dev2 , fixed via #96915 .

I'm closing the issue since this is becoming misleading, and as mentioned above, the OP issue was that the 2 projects had different functions, and the engine was correctly reporting what the issue was:

```
ERROR: The rpc node checksum failed. Make sure to have the same methods on both nodes. Node path: Main
   at: process_simplify_path (modules/multiplayer/scene_cache_interface.cpp:121)
```
